### ISSUE TITLE
Why are trainset, valset, testset the same?

### ISSUE THREAD

**Issue by jinjiao97** [lxd99/CTCP#2]
line1: https://github.com/lxd99/CTCP/blob/d1c4d0ee2c7ab8b42cd2ce58e39029600d440159/train/train.py#L64C49-L64C49
line2: https://github.com/lxd99/CTCP/blob/d1c4d0ee2c7ab8b42cd2ce58e39029600d440159/train/train.py#L87C38-L87C38
Dear author,
You consider the trainset, valset and testset which are the same( line1). In other words, you take the retweet information of the testset and valset into training. Although your loss backward process is under the situation where a cascade reaches the observation time, the cumulative gradient is from the last cascade to the current cascade where some retweets are from testset or valset. Is it not an information leakage？If the retweets from testset or valset is zero_grad , I think it's ok. But zero_grad operation is from each batch where a cascade reaches the observation time( line 2). Every gradient descent may be from some data which is valset or testset. Could you please explain the rationality for this?

Sincerely

**Comment by lxd99** [comment]
Thanks for your question. Please note that, in the training process, although the model can see the retweet of validating and testing cascades, the labels (i.e., popularity) of these cascades are not leakage since we only use the training cascades to compute the loss (shown in the following code).

https://github.com/lxd99/CTCP/blob/d1c4d0ee2c7ab8b42cd2ce58e39029600d440159/train/train.py#L79

For the rationality of this setting, the cascade popularity prediction can be regarded as a node regression problem on a dynamic graph, and such an experimental setting is an analogy to the widely used setting of [node classification task](https://docs.dgl.ai/tutorials/blitz/1_introduction.html#sphx-glr-tutorials-blitz-1-introduction-py) on the static graph, where the mode can see the features of all nodes while only the label of training node is used to compute the loss.

**Comment by jinjiao97** [comment]
> Thanks for your question. Please note that, in the training process, although the model can see the retweet of validating and testing cascades, the labels (i.e., popularity) of these cascades are not leakage since we only use the training cascades to compute the loss (shown in the following code).
> 
> https://github.com/lxd99/CTCP/blob/d1c4d0ee2c7ab8b42cd2ce58e39029600d440159/train/train.py#L79
> 
> For the rationality of this setting, the cascade popularity prediction can be regarded as a node regression problem on a dynamic graph, and such an experimental setting is an analogy to the widely used setting of [node classification task](https://docs.dgl.ai/tutorials/blitz/1_introduction.html#sphx-glr-tutorials-blitz-1-introduction-py) on the static graph, where the mode can see the features of all nodes while only the label of training node is used to compute the loss.

Thanks for your reply. I don't mean the leakage of the label. I mean the leakage of retweet information in testing and validating cascades. For example, in batch data, user 'b' retweets a testing cascade from user 'a' while user 'c' retweets a training cascade from user 'a'. However, the information where user 'b' retweets a cascade from user 'a' should not be used in the training process because it is a testing cascade. Is it not a leakage?

As for the setting where only the label of the training node is used to compute the loss, I think it is totally reasonable. My key point is whether it is suitable to add the testing retweets behavior to the training process.

**Comment by lxd99** [comment]
From my point of view, the key point of information leakage is whether the label of testing data is seen by the model rather than the input of testing data. In a real-world machine learning scenario, we can actually get the input of training data, the label of training data, the input of evaluating data, the label of evaluating data, and the input of testing data, the only unknown is the label of testing data.  So, if the training process uses the input of testing data, it does not conflict with the real-world scenario. However, your concern is also reasonable, and removing the testing retweets from the training set might be a more rigorous setting.

**Comment by subetter** [comment]
Dear author,
In your work you have compared a variety of advanced methods, but why is the data in your table different from the data presented in the original paper under the same Settings?  Can you explain how you got the data for your table?
Sincerely

**Comment by lxd99** [comment]
@subetter  Thanks for your question. Please note that, although under the same experimental setting, the performance of a model can vary due to differences in the environment (e.g., python package version and hardware). Specifically, throughout the experiment, certain operations involve randomness, and these variations in the environment can lead to divergent results and different model performances. We list some of such operations as follows.

1) The selection of the training, validating, and testing cascades is random. Our selection may not be the same as the original paper.
2) The model initialization is random, our initialization may not be the same as the original paper.
3) The result of the dropout layer is random.
